rfc/006-backfill-blocks.md

```diff
@@ -0,0 +1,187 @@
+to meet the retention limit. This could be a problem as nodes who state sync and
+then eventually switch to consensus may not have the block and validator
+history to verify evidence causing them to panic if they see 2/3 commit on what
+the node believes to be an invalid block.
+
+Having the capability to backfill prior blocks also allows nodes with
+truncated history to become archive nodes which may be beneficial for the
+network.                                            
+
+## Proposal
+
+A backfill mechanism can simply be defined as an algorithm for fetching,
+verifying and storing, blocks and other state data (`ValidatorSet`'s
+`ConsensusParams`'s and `ABCIResponses`'s) of a height prior to the current
+base of the blockchain. Verification and design will be addressed
+later in the RFC. First we cover where the mechanism would be deployed.
+
+There are two places where a backfill mechanism would be required:
+1. Upon startup via state sync where nodes must meet the prerequisite history
+before participating in consensus
+2. Where application specifies that the `retain_height` needs to be lower than
+the current blockchain base.
+
+### Backfill on State Sync
+
+A node, seeking to state sync, will find an adequate snapshot and offer it to
+the application with the following format:
+
+```proto
+message RequestOfferSnapshot {
+  Snapshot snapshot = 1;  // snapshot offered by peers
+  bytes    app_hash = 2;  // light client-verified app hash for snapshot height
+}
+```
+
+where `Snapshot` contains the new field `backfill_height` :
+
+```proto
+message Snapshot {
+  uint64 height          = 1;  // The height at which the snapshot was taken
+  uint32 format          = 2;  // The application-specific snapshot format
+  uint32 chunks          = 3;  // Number of chunks in the snapshot
+  bytes  hash            = 4;  // Arbitrary snapshot hash, equal only if identical
+  bytes  metadata        = 5;  // Arbitrary application metadata
+  int64  backfill_height = 6;  // Height to backfill blocks from before starting application (inclusive)
+}
+```
+
+If all chunks are accepted and state sync is successful then the node will
+retrieve and verify blocks up to (and including) the specified backfill height
+**before** participating in consensus.
+
+### Backfill on application request
+
+The application has control of block retention via `retain_height`, called here:
+
+message ResponseCommit {
+  // reserve 1
+  bytes data          = 2;
+  int64 retain_height = 3;
+}
+
+Beforehand, a retain height that was less than the nodes current base was
+ignored. With the same backfill mechanism, the application should now be able to
+lower the retention height below the current base height (whether to accommodate
+changing consensus parameters, app specific needs, or a local config).
+This accommodates the use case of a truncated node that wishes to become an
+archive node. Note this could be done simultaneously alongside consensus.
```

Comment:
> How do we determine whether backfill is synchronous or asynchronous?

I'm guessing the node will determine this i.e. if it's during startup after state sync then we need to do this before consensus. Outside of startup we can run it asynchronously so long as the implementation permits it (we may not be able to drop back into fast sync if we are already backfilling).

> There's a footgun here: if the application returns retain_height: 10 on all nodes, such that all nodes delete 1-9, and then returns retain_height: 5 later, backfill will be impossible because no nodes have those blocks.

Yeah good call. It also might be that the blocks still exist but none of our peers have it. I guess we sort of imply (or hope) that applications aren't going to shoot themselves in the foot by doing this. Outside of this we can do the same thing as fastsync where we query the lowest height that our peers have and if it's greater than the retain height then we just terminate the service once we reach that height.